File: src/main/java/com/github/kindrat/cassandra/client/service/tasks/AbstractTask.java
package com.github.kindrat.cassandra.client.service.tasks;

import lombok.extern.slf4j.Slf4j;
import reactor.core.Disposable;
import reactor.core.publisher.EmitterProcessor;
import reactor.core.publisher.Flux;

import java.util.Optional;

@Slf4j
public abstract class AbstractTask implements BackgroundTask {
    private final EmitterProcessor<TaskStatistics> eventStream = EmitterProcessor.create(false);
    private TaskStatistics lastState;
    private Disposable job;

    @Override
    public TaskStatistics getStatistics() {
        return lastState;
    }

    @Override
    public Flux<TaskStatistics> run() {
        job = innerRun().subscribe(eventStream::onNext, eventStream::onError);
        return eventStream;
    }

    abstract Flux<TaskStatistics> innerRun();

    @Override
    public void dispose() {
        Optional.ofNullable(job).ifPresent(Disposable::dispose);
        lastState = lastState.toBuilder()
                .state(TaskState.DISCARDED)
                .build();
    }

    @Override
    public boolean isDisposed() {
        return lastState.getState() != TaskState.ACTIVE;
    }
}

File: src/main/kotlin/com/github/kindrat/cassandra/client/service/tasks/AbstractTask.kt
package com.github.kindrat.cassandra.client.service.tasks

import com.github.kindrat.cassandra.client.service.tasks.TaskState.ACTIVE
import reactor.core.Disposable
import reactor.core.publisher.EmitterProcessor
import reactor.core.publisher.Flux
import java.lang.System.currentTimeMillis
import java.time.Duration
import java.util.*

abstract class AbstractTask : BackgroundTask {
    private val eventStream = EmitterProcessor.create<TaskStatistics>(false)

    override var statistics: TaskStatistics = TaskStatistics(ACTIVE, currentTimeMillis(), Duration.ZERO, 0.0)

    private var job: Disposable? = null
    override fun run(): Flux<TaskStatistics> {
        job = innerRun().subscribe({ eventStream.onNext(it) }) { eventStream.onError(it) }
        return eventStream
    }

    abstract fun innerRun(): Flux<TaskStatistics>

    override fun dispose() {
        job?.apply { this.dispose() }
        statistics = statistics.copy(state = TaskState.DISCARDED)
    }

    override fun isDisposed(): Boolean {
        return statistics.state !== ACTIVE
    }
}